### ISSUE TITLE
Fail to load pre-trained tokens.

### ISSUE THREAD

**Issue by xijiz** [huggingface/transformers#878]
The PreTrainedTokenizer fails to load tokenizer files when I load tokenizer files from local tokenizer files.

The error is caused by code line 174 - 182 in [tokenization_utils.py](https://github.com/huggingface/pytorch-transformers/blob/master/pytorch_transformers/tokenization_utils.py). The code assumes that there are three tokenizer files: added_tokens.json, special_tokens_map.json, vocab.txt. However, these path of tokenizer files will be the same path if the given parameter "pretrained_model_name_or_path" is full path of "vocab.txt".

**Comment by bbruceyuan** [comment]
hi, I meet the same problem too.

have you solved it? if so, please show me your way. 

additionally, I have another questions.
Q. where can I download the `BertTokenizer.from_pretrained(pretrained_model_name_or_path)`?
  I mean what's the url of `pretrained_model_name_or_path` that I can download previously.

thanks in advance!

**Comment by xijiz** [comment]
------------------ 原始邮件 ------------------
发件人: "Chaofa Yuan"<notifications@github.com>;
发送时间: 2019年8月1日(星期四) 晚上9:00
收件人: "huggingface/pytorch-transformers"<pytorch-transformers@noreply.github.com>;
抄送: "曾祥极"<xijiz@qq.com>;"Author"<author@noreply.github.com>;
主题: Re: [huggingface/pytorch-transformers] Fail to load pre-trainedtokens. (#878)




hi, I meet the same problem too.
 
have you solved it? if so, please show me your way.
 
additionally, I have another questions.
 Q. where can I download the BertTokenizer.from_pretrained(pretrained_model_name_or_path)?
 I mean what's the url of pretrained_model_name_or_path that I can download previously.
 
thanks in advance!
 
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.




I solved it by previously downloading BertTokenizer and pre-trained BertModel. You can download these files by urls that can be found in model code files.

- bert:

https://github.com/huggingface/pytorch-transformers/blob/master/pytorch_transformers/modeling_bert.py

BERT_PRETRAINED_MODEL_ARCHIVE_MAP = {
    'bert-base-uncased': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-uncased-pytorch_model.bin",
    'bert-large-uncased': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-large-uncased-pytorch_model.bin",
    'bert-base-cased': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-cased-pytorch_model.bin",
    'bert-large-cased': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-large-cased-pytorch_model.bin",
    'bert-base-multilingual-uncased': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-multilingual-uncased-pytorch_model.bin",
    'bert-base-multilingual-cased': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-multilingual-cased-pytorch_model.bin",
    'bert-base-chinese': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-chinese-pytorch_model.bin",
    'bert-base-german-cased': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-german-cased-pytorch_model.bin",
    'bert-large-uncased-whole-word-masking': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-large-uncased-whole-word-masking-pytorch_model.bin",
    'bert-large-cased-whole-word-masking': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-large-cased-whole-word-masking-pytorch_model.bin",
    'bert-large-uncased-whole-word-masking-finetuned-squad': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-large-uncased-whole-word-masking-finetuned-squad-pytorch_model.bin",
    'bert-large-cased-whole-word-masking-finetuned-squad': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-large-cased-whole-word-masking-finetuned-squad-pytorch_model.bin",
    'bert-base-cased-finetuned-mrpc': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-cased-finetuned-mrpc-pytorch_model.bin",
}
 
BERT_PRETRAINED_CONFIG_ARCHIVE_MAP = {
    'bert-base-uncased': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-uncased-config.json",
    'bert-large-uncased': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-large-uncased-config.json",
    'bert-base-cased': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-cased-config.json",
    'bert-large-cased': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-large-cased-config.json",
    'bert-base-multilingual-uncased': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-multilingual-uncased-config.json",
    'bert-base-multilingual-cased': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-multilingual-cased-config.json",
    'bert-base-chinese': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-chinese-config.json",
    'bert-base-german-cased': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-german-cased-config.json",
    'bert-large-uncased-whole-word-masking': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-large-uncased-whole-word-masking-config.json",
    'bert-large-cased-whole-word-masking': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-large-cased-whole-word-masking-config.json",
    'bert-large-uncased-whole-word-masking-finetuned-squad': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-large-uncased-whole-word-masking-finetuned-squad-config.json",
    'bert-large-cased-whole-word-masking-finetuned-squad': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-large-cased-whole-word-masking-finetuned-squad-config.json",
    'bert-base-cased-finetuned-mrpc': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-cased-finetuned-mrpc-config.json",
}

https://github.com/huggingface/pytorch-transformers/blob/master/pytorch_transformers/tokenization_bert.py
PRETRAINED_VOCAB_FILES_MAP = {
    'vocab_file':
    {
        'bert-base-uncased': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-uncased-vocab.txt",
        'bert-large-uncased': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-large-uncased-vocab.txt",
        'bert-base-cased': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-cased-vocab.txt",
        'bert-large-cased': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-large-cased-vocab.txt",
        'bert-base-multilingual-uncased': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-multilingual-uncased-vocab.txt",
        'bert-base-multilingual-cased': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-multilingual-cased-vocab.txt",
        'bert-base-chinese': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-chinese-vocab.txt",
        'bert-base-german-cased': "https://int-deepset-models-bert.s3.eu-central-1.amazonaws.com/pytorch/bert-base-german-cased-vocab.txt",
        'bert-large-uncased-whole-word-masking': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-large-uncased-whole-word-masking-vocab.txt",
        'bert-large-cased-whole-word-masking': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-large-cased-whole-word-masking-vocab.txt",
        'bert-large-uncased-whole-word-masking-finetuned-squad': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-large-uncased-whole-word-masking-finetuned-squad-vocab.txt",
        'bert-large-cased-whole-word-masking-finetuned-squad': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-large-cased-whole-word-masking-finetuned-squad-vocab.txt",
        'bert-base-cased-finetuned-mrpc': "https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-cased-finetuned-mrpc-vocab.txt",
    }
}

**Comment by jessica-lll** [comment]
Hi, how do you solve this problem? If we set `pretrained_model_name_or_path` as a path to vocab.txt, it still need the two files: added_tokens.json, special_tokens_map.json. Where can we get these files?

**Comment by xijiz** [comment]
> Hi, how do you solve this problem? If we set `pretrained_model_name_or_path` as a path to vocab.txt, it still need the two files: added_tokens.json, special_tokens_map.json. Where can we get these files?

You can ignore these files: added_tokens.json, special_tokens_map.json. All you need to do is to modify some code lines in the file: tokenization_utils.py. I have modified it in my forked repository as you can see [here](https://github.com/xijiz/pytorch-transformers/commit/31bc1ddf4f68ad790da9874a3623cf22d62dc186).
